### PR TITLE
Fix unexpected parts handling

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Follow-up to https://github.com/ClickHouse/ClickHouse/pull/56142

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit a4b32d806aa41c43986f8b9028b2659929fa8185 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56693/a4b32d806aa41c43986f8b9028b2659929fa8185/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6854327033">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56693/a4b32d806aa41c43986f8b9028b2659929fa8185/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -1422,6 +1435,16 @@ void StorageReplicatedMergeTree::checkParts(bool skip_sanity_checks)
             continue;
         }
 
+        {
+            std::lock_guard lock(outdated_data_parts_mutex);
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:1439]: It's very unclear that we are going to retry only when we have unexpected uncovered parts. Add a comment.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -1327,6 +1327,19 @@ void StorageReplicatedMergeTree::paranoidCheckForCoveredPartsInZooKeeperOnStart(
 }
 
 void StorageReplicatedMergeTree::checkParts(bool skip_sanity_checks)
+{
+    if (checkPartsImpl(skip_sanity_checks))
+        return;
+
+    waitForOutdatedPartsToBeLoaded();
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:1334]: Add a comment

**Review by alesapin — APPROVED** [review]

**Comment by tavplubix** [comment]
Integration tests (tsan) [5/6] - https://github.com/ClickHouse/ClickHouse/pull/56796
Performance Comparison [3/4] - https://github.com/ClickHouse/ClickHouse/pull/56797
Stateless tests (debug) [3/5] - #56685
Stress test (msan) - https://github.com/ClickHouse/ClickHouse/issues/56798

**Pull request merged**